blog/_posts/2016-02-01-iteration.md

```diff
@@ -0,0 +1,420 @@
+```
+
+You can construct these manually: for example,
+
+```jl
+julia> CartesianRange(CartesianIndex((-7,0)), CartesianIndex((7,15)))
+CartesianRange{CartesianIndex{2}}(CartesianIndex{2}((-7,0)),CartesianIndex{2}((7,15)))
+```
+
+constructs a range that will loop over `-7:7` along the first
+dimension and `0:15` along the second.
+
+One reason that `eachindex` is recommended over `for i = 1:length(A)`
+is that some `AbstractArray`s cannot be indexed efficiently with a
+linear index; in contrast, a much wider class of objects can be
+efficiently indexed with a multidimensional iterator.  (SubArrays are,
+generally speaking, a prime example.)  `eachindex` is designed to pick
+the most efficient iterator for the given array type.  You can even
+use
+
+```jl
+for i in eachindex(A, B)
+...
+```
+
+to ensure that `i` will be efficient for accessing both `A` and `B`.
+
+As we'll see below, these iterators have another purpose: independent
+of whether the underlying arrays have efficient linear indexing,
+multidimensional iteration can be a powerful ally when writing
+algorithms.  The rest of this blog post will focus on this
+latter application.
+
+## Writing multidimensional algorithms with `CartesianIndex` iterators
+
+### A multidimensional boxcar filter
+
+Let's suppose we have a multidimensional array `A`, and we want to
+compute the ["moving
+average"](https://en.wikipedia.org/wiki/Boxcar_averager) over a
+3-by-3-by-... block around each element.  From any given index position,
+we'll want to sum over a region offset by `-1:1` along each dimension.
+Edge positions have to be treated specially, of course, to avoid going
+beyond the bounds of the array.
+
+In many languages, writing a general (N-dimensional) implementation of
+this conceptually-simple algorithm is somewhat painful, but in Julia
+it's a piece of cake:
+
+```jl
+function boxcar3{T,N}(A::AbstractArray{T,N})
+    out = similar(A)
+    I1 = CartesianIndex(ntuple(d->1, N)::NTuple{N,Int})
+    Iend = CartesianIndex(size(A))
+    for I in CartesianRange(size(A))
+        n = 0
+        s = zero(T)
+        for J in CartesianRange(max(I1, I-I1), min(Iend, I+I1))
+            s += A[J]
+            n += 1
+        end
+        out[I] = s/n
+    end
+    out
+end
+```
+
+Let's walk through this line by line:
+
+- `out = similar(A)` allocates the output. In a "real" implementation,
+  you'd want to be a little more careful about the element type of the
+  output (what if the input array element type `T` is `Int`?), but
+  we're cutting a few corners here for simplicity.
+
+- `I1 = CartesianIndex(ntuple(d->1, N)::NTuple{N,Int})` creates an
+  `N`-dimensional `CartesianIndex` filled with 1s, e.g.,
+  `CartesianIndex{3}(1,1,1)` in three dimensions.  The type
+  annotation, `::NTuple{N,Int}` is unfortunately necessary (if you
+  want good performance), because julia 0.4 cannot infer the output
+  type of (most) calls to `ntuple`.  We'll explore this topic further
+  below.
```

Comment:
Great idea about using `first` and `last` here. Being able to use `first` to compute the offset range is, of course, a trick that's very specific to the fact that we're doing 3-by-3-by... (it wouldn't work if it were 5-by-5-by...). But it **is** the best way to handle the enforcement of the bounds, so in that respect this is an improvement in terms of generality.

As you say, it would also be better to avoid introducing type instability so early, so for pedagogical reasons I'll take all the dirty tricks I can get! Thanks!